File: Nodejs/merge/merge.js
var fs = require('fs'),
	readline = require('readline'),
	LineReaderSync = require("line-reader-sync");

const ENV = "dev";
const rootpath = "resources/mergesort/";

var print = (function () {
    return {
        log: function() {
            var args = Array.prototype.slice.call(arguments);
            if (ENV === 'dev') {
            	console.log.apply(console, args);
            }
        }
    }
}());

var GetLeaves = function(dirpath) {
	return fs.readdirSync(rootpath + dirpath);
}

var MergeSort = function(dirpath,cb) {
	var leaves = GetLeaves(dirpath);
	let sortedpath = dirpath + "sorted/";
	createFolder(sortedpath);
	for (var i in leaves) {
		// print.log("sort MergeSort: ",dirpath + leaves[i]);
		SortFile(dirpath + leaves[i]);
	}
	leaves = GetLeaves(sortedpath);
	// print.log("leaves: ",leaves, "length: ",leaves.length);
	mergesort(sortedpath,leaves,0);
	cb(null);
}

var mergesort = function(path,m,cont){
	if (m.length <= 1) {
		return m;
	}

	print.log("m.length: ",m.length);

	let mid = m.length / 2;
	let left = m.slice(0,mid);
	let right = m.slice(mid,m.length);

	print.log("mid: " + mid);
	print.log("left arr: " + left);
	print.log("right arr: " + right);

	left = mergesort(path,left,cont+1);
	right = mergesort(path,right,cont+2);

	return merge(path,left, right,cont);
}

var merge = function(path,left, right, cont) {
	var result = [];
	let name = cont + "_" + (left[0]).split(".")[0] + "_" + (right[0]).split(".")[0] + ".merged.sorted";
	if (cont == 0) {name = cont + ".merged.sorted";};
	print.log("sorted name: ",name," left: ",left," right: ",right);
	result.push(name);
	let left_file = new LineReaderSync(rootpath + path + left[0]);
	let right_file = new LineReaderSync(rootpath + path + right[0]);
	let new_file_path = rootpath + path + name;

	let left_word = left_file.readline();
	let right_word = right_file.readline();
	while(true) {
		if (left_word == null) {
			left_word = left_file.readline();
		}
		if (right_word == null) {
			right_word = right_file.readline();
		}

		if (left_word != null && right_word != null) {
			print.log("left: "+left_word + " " + left_word.length);
			print.log(" right: "+right_word+ " " + right_word.length);
			if (left_word <= right_word) {
				fs.appendFileSync(new_file_path,left_word+"\n");
				left_word = null;
			} else {
				fs.appendFileSync(new_file_path,right_word+"\n");
				right_word = null;
			}
		} else if (left_word != null) {
			print.log("lesf: "+left_word + " " + left_word.length);
			fs.appendFileSync(new_file_path,left_word+"\n");
			left_word = null;
		} else if (right_word != null) {
			print.log(" right: "+right_word+ " " + right_word.length);
			fs.appendFileSync(new_file_path,right_word+"\n");
			right_word = null;
		}else{
			break;
		}
	}

	return result
}

var createFolder = function(filename) {
	let folder = filename.split(".")[0];
	fs.mkdirSync(rootpath+folder);
}

var FilterFile = function(filepath, pattern,cb) {
	let regx = new RegExp(pattern);
	let file = fs.openSync(rootpath + filepath,'r');
	
	let createdFile = rootpath +filepath + ".filtered";
	
	// print.log("Readfile: "+rootpath + filepath);
	// print.log("filtered: "+createdFile);
	let rd = readline.createInterface({
	    input: fs.createReadStream(rootpath + filepath),
	    output: process.stdout,
	    terminal: false
	});

	rd.on('line',(line) => {
		if (match = regx.exec(line)) {
			// print.log(line, filteredFile);
			fs.appendFileSync(createdFile,line.toLowerCase()+"\n");
		};
	});

	fs.closeSync(file);

	rd.on('close',function(){
		cb(null);
	});
}

var CreateLeaves = function(filepath, leafSize, cb) {
	let folder = rootpath + filepath.split("/")[0] + "/leaves/";
	createFolder(folder.replace(rootpath,""));
	let leafcont = 0;
	var filteredFile;
	// print.log("path: " + rootpath + filepath);
	var rd = readline.createInterface({
	    input: fs.createReadStream(rootpath + filepath),
	    output: process.stdout,
	    terminal: false
	});

	var linecont = 0;
	rd.on('line',(line) => {
		if (linecont%leafSize == 0) {
			if(linecont>0)
				fs.closeSync(filteredFile);
			// print.log(linecont);
			filteredFile = fs.openSync(folder +"leaf" + leafcont,"a+");
			leafcont++;
		}
		fs.writeSync(filteredFile,line+"\n");
		linecont++;
	});

	rd.on('close', function(){
		cb(null);
	});
}

var SortFile = function(filepath) {
	let fileData = fs.readFileSync(rootpath + filepath,'utf8');
	let data = fileData.split("\n");
	data.sort((a,b) => {
		if (a < b)
			return -1;
		else if (a === b) 
			return 0;
		else
			return 1;
	});
	let fullpath = filepath.split("/");
	let filename_temp = fullpath.pop();
	fullpath.push("sorted");
	// createFolder(fullpath.join('/'));
	fullpath.push(filename_temp);
	// print.log("fullpath: "+fullpath);
	// print.log("swap fullpath: "+fullpath);
	sortedpath = rootpath + fullpath.join('/');
	// print.log("path: "+ sortedpath);
	let sortedFile = fs.openSync(sortedpath + ".sorted", "a+");
	data = data.slice(1,data.length) //delete '\n'
	for (var line in data) {
		fs.writeSync(sortedFile,data[line]+"\n");
	}
	fs.closeSync(sortedFile);
}

var GetSortedFile = function(name){
	return rootpath + name.split(".")[0] +"/leaves/sorted/0.merged.sorted";
}

// var files = 
// SortFile("emails/emails.txt");
// print.log(files);

var test = function(){
	let filename = "emails.txt";
	let name = filename.split(".")[0];
	FilterFile(name+"/"+filename,"(\\w[-._\\w]*\\w@\\w[-._\\w]*\\w\\.\\w{2,3})",function(err){
		if (err) {
			print.log("error filter");
		}else{
			print.log("termino filter");
			CreateLeaves(name+"/"+filename+".filtered",5, function(err){
				if(err)
					print.log("error create leaves");
				else{
						print.log("termino filter");
						MergeSort(name+"/leaves/",function(err){
							if (err) {
								print.log("error MergeSort");
							}
						});
					}
			});
		}
	});
}

exports.FilterFile = FilterFile;
exports.CreateLeaves = CreateLeaves;
exports.MergeSort = MergeSort;
exports.GetSortedFile = GetSortedFile;

// test();

// FilterFile("emails/emails.txt",/^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i);

File: Ruby/merge/merge.rb
require 'fileutils'
$rootpath_sort = "resources/mergesort/"

def GetLeaves(dirpath)
	files_name = Dir.entries("#{$rootpath_sort}#{dirpath}")
	
	puts "#{dirpath}"
	files_name.delete(".")
	files_name.delete("..")
	puts files_name
	files_name
end

def MergeSort(dirpath)
	leaves = GetLeaves(dirpath)
	sortedpath = "#{dirpath}sorted/"
	createFolder(sortedpath)
	for i in (0..(leaves.length-1))
		SortFile("#{dirpath}#{leaves[i]}")
	end

	leaves = GetLeaves(sortedpath)
	mergesort(sortedpath,leaves,0,"")
end

def mergesort(path,m,cont,path_to)
	if m.length <= 1
		return m
	end

	mid = m.length / 2
	left = m[0,mid]
	right = m[mid,m.length]

	left = mergesort(path,left,cont+1,"#{path_to}#{cont}")
	right = mergesort(path,right,cont+2,"#{path_to}#{cont}")

	return merge(path,left, right,cont,path_to)
end

def merge(path,left, right, cont,path_to)
	result = []
	# name = "#{cont}.merged.sorted"
	# unless cont == 0
	# 	# name = "#{cont}_#{(left[0]).split(".")[0]}_#{(right[0]).split(".")[0]}.merged.sorted"
	# 	name = "#{cont}_#{(left[0]).split(".")[0]}_#{(right[0]).split(".")[0]}.merged.sorted"
	# end
	name = "#{path_to}#{cont}.merged.sorted"
	result.push(name)
	left_file = File.open("#{$rootpath_sort}#{path}#{left[0]}",'r')
	right_file = File.open("#{$rootpath_sort}#{path}#{right[0]}",'r')
	new_file = File.open("#{$rootpath_sort}#{path}#{name}",'w')

	pos_left = 0
	pos_right = 0

	left_word = nil
	right_word = nil
	while true
		if left_word == nil
			left_word = left_file.gets
		end
		if right_word == nil
			right_word = right_file.gets
		end

		if left_word != nil and right_word != nil
			puts "left: #{left_word} #{-left_word.length}"
			puts " right: #{right_word} #{-right_word.length}"
			if left_word <= right_word
				new_file.write(left_word)
				left_word = nil
			else
				new_file.write(right_word)
				right_word = nil
			end
		elsif left_word != nil
			puts "left: #{left_word} #{-left_word.length}"
			new_file.write(left_word)
			left_word = nil
		elsif right_word != nil
			puts " right: #{right_word} #{-right_word.length}"
			new_file.write(right_word)
			right_word = nil
		else
			break
		end
	end

	left_file.close()
	right_file.close()
	new_file.close()

	return result
end

def createFolder(filename)
	folder = filename.split(".")[0]
	Dir.mkdir("#{$rootpath_sort}#{folder}")
end

def FilterFile(filepath, pattern)
	filteredFile = File.open("#{$rootpath_sort}#{filepath}.filtered",'w')
	# filteredFile = File.open("resources/bitcode/2milmails/2milmails.txt.filtered",'w')
	f = File.open("#{$rootpath_sort}#{filepath}",'r')
	while(line = f.gets)
		puts "entro"
		puts line
		begin
			if(line =~ pattern)
				filteredFile.write("#{line.downcase}\n")
				puts line
				puts "dentro if"
			elsif
				unless line.nil?
					raise line
				end
			end
		rescue
			line = line[0,line.length-1]
			retry
		end
		puts "paso"
	end
	f.close
	filteredFile.close
end

def CreateLeaves(filepath, leafSize)
	folder = "#{$rootpath_sort}#{filepath.split("/")[0]}/leaves/"
	createFolder("#{filepath.split("/")[0]}/leaves/")
	file = File.open("#{$rootpath_sort}#{filepath}",'r')
	
	leafcont = 0
	linecont = 0
	filteredFile = ""

	while(line = file.gets)
		if(linecont%leafSize == 0)
			if(linecont > 0)
				filteredFile.close
			end
			filteredFile = File.open("#{folder}leaf#{leafcont}",'a')
			leafcont += 1
		end
		filteredFile.write(line)
		linecont = linecont + 1
	end

	file.close
	filteredFile.close
end

def SortFile(filepath)
	File.open("#{$rootpath_sort}#{filepath}",'r') do |f|
		fileData = f.readlines
		data = fileData.sort

		print data
		
		puts "filepath:", filepath
		fullpath = filepath.split("/"); 
		puts "fullpath:", fullpath
		fullpath.push("sorted")
		l = fullpath.length
		fullpath[l-2],fullpath[l-1] = fullpath[l-1], fullpath[l-2]
		sortedpath = fullpath.join("/")
		
		puts "path: "+ sortedpath
		File.open("#{$rootpath_sort}#{sortedpath}.sorted",'w') do |sortedFile|
			for i in 0..(data.length-1)
				sortedFile.write(data[i])
			end
		end
	end
end

def GetSortedFile(name)
	"#{$rootpath_sort}#{name.split('.')[0]}/leaves/sorted/0.merged.sorted"
end